### DISCUSSION TITLE
Refinement on objects executed only after other fields have valid type

### DISCUSSION THREAD

**Discussion by bartosjiri** [colinhacks/zod#1914] [Q&A] [answered]
Having an object with refinement, it seems that the refinement is executed only after some other fields have valid type.

Example:
```ts
z.object({
  title: z.string().min(3).max(128),
  minLength: z.number().min(1).optional().default(1),
  maxLength: z.number().max(512).optional().default(512)
}).refine(({minLength, maxLength}) => maxLength >= minLength, {
  message: "Must be smaller than or equal to minimum",
  path: ["maxLength"]
})
```

When I run `safeParse` after providing for example `300` to `minLength` and `200` to `maxLength`, leaving the `title` empty, I get errors only for the `title` field -> `title: Expected string, received null`. 

Only after the type for `title` is valid (but can still fail `min` or `max` constraint), I get the desired error for the refinement. Following the step above and providing single character value to the `title` field -> `title: String must contain at least 3 character(s); maxLength: Must be smaller than minimum`.

Seems closely related (if not the same) to https://github.com/colinhacks/zod/discussions/437 and the issue referenced in it.
Is that the case or is there any way to execute refinements even with invalid input type on "unrelated" fields?

**Comment by JacobWeisenburger** [comment, marked as answer]
Does this work for you?

```ts
const baseSchema = z.object( {
    title: z.string().min( 3 ).max( 128 ),
    minLength: z.number().min( 1 ).optional().default( 1 ),
    maxLength: z.number().max( 512 ).optional().default( 512 )
} )

const schema = baseSchema.pick( { minLength: true, maxLength: true } )
    .passthrough()
    .refine( ( { minLength, maxLength } ) => maxLength >= minLength, {
        message: 'Must be smaller than or equal to minimum',
        path: [ 'maxLength' ]
    } )
    .pipe( baseSchema )

const passResult = schema.safeParse( {
    title: 'test',
    minLength: 150,
    maxLength: 200,
} )
passResult.success && console.log( passResult.data )
// { title: 'test', minLength: 150, maxLength: 200 }

const failResult = schema.safeParse( { minLength: 300, maxLength: 200 } )
!failResult.success && console.log( failResult.error.issues )
// [
//     {
//         code: 'custom',
//         message: 'Must be smaller than or equal to minimum',
//         path: [ 'maxLength' ]
//     }
// ]
```

**Reply by bartosjiri** [reply]
It almost does work - I believe now the `title` does not get errors until the refinement is valid. 
But I'll be able to figure that out, thanks for showing me the direction, appreciate your time and help!
